### DISCUSSION TITLE
Questions and ideas about Ki

### DISCUSSION THREAD

**Discussion by C-BJ** [ki-lang/ki#18] [General]
Some questions about Ki:
1. Functional programming?
2. OOP?
3. meta, generic?
4. Self-hosted?
5. How to allocate memory from the heap (common variables? Class? Struct?)
6. custom memory allocator?
7. when shallow copy, deep copy?
8. embedded?
9. Ideas for language design? Simplicity? Orthogonal? Simplicity? Robust? Readability? Writability? Powerful?
10. Idea of package manager?
11. Scope of the standard library(What is covered by the Standard Library Program)?
12. Scope of RC, Ownership(When can I not use them?)?
13. Hidden control flow?
14. How is Map implemented?
15. Hash Map?
16. Pattern matching?
17. Function overloading? (I think either none, or use an explicit function overload like Odin's  https://odin-lang.org/docs/overview/#explicit-procedure-overloading)
18. Can I set a function in a file to be private?
19. Can I create a function or struc/class in function
20. for?
21. What is the syntax of inline function?

-----------------------------------
Some ideas for Ki (maybe some Ki already have):
1. borrowing from Jai/Odin syntax(Especially Odin)?
2. Semicolon? (I know this is already a question under consideration)
3. let the function return two variables at the same time and assign values at the call?
```
//Odin
swap :: proc(x, y: int) -> (int, int) {
    return y, x
}
a, b := swap(1, 2)
fmt.println(a, b) // << 2 1
```
4. something like Zig's break / continue https://github.com/ratfactor/ziglings/blob/268e8963666723f96d8cd78ced61c6ad0c047037/exercises/063_labels.zig#LL1C1-L52C3
5. Take inspiration from cppfront?
https://github.com/hsutter/cppfront
6. Create an error set with multiple errors?
```
let message_error = {!message_too_long, !message_too_short}
func hello(message: *String) void !message_error  {
  if(message.bytes() < 10) : throw message_error.message_too_short;
  if(message.bytes() > 200) : throw message_error.message_too_long;
  println(message);
}
```
7. use CPP lib?
8. defer?
9. anonymous blocks?
```
func main() void {
  {
    let x = 1;
    println(x.str()); // << 1
  }
  let x = 2;
  println(x.str()); // << 2
}
```
10. The syntax of the declaration?
https://odin.handmade.network/blog/p/2994-on_the_aesthetics_of_the_syntax_of_declarations
11. A random thought of my own. sth like:
```
fn main() i32 = { }
class animal = { };
let a = 1
```
```const main = fn() i32 { }
const animal = class{ };
let a = 1
```
```
main::fn() i32 { }
animal::class{ };
a := 1 // mut
b :: 2 //const
```
12. named and optional arguments?

**Comment by ctxcode** [comment]
1. Functional programming?
> no
2. OOP?
> no
3. meta, generic?
> meta: maybe (but keep it minimal), generic: yes
4. Self-hosted?
> maybe, it's extra work for no reason
5. How to allocate memory from the heap (common variables? Class? Struct?)
> `use ki:mem;  mem:alloc({size});`
6. custom memory allocator?
> maybe, you can create your own allocator, but to have it built in into language, we'll see. It's nice to have, but it has low prio for now.
7. when shallow copy, deep copy?
> The compiler should generate copy functions, but it doesnt right now. I think it's very useful.
8. embedded?
> I dont know what makes a language embedded, you mean like `ki.h` & `ki.a` so other programs can link against it? Maybe way in the future.
9. Ideas for language design? Simplicity? Orthogonal? Simplicity? Robust? Readability? Writability? Powerful?
> These are just buzz words, but simplicity is very important
10. Idea of package manager?
> We already have this, we just need to create a platform to share packages
11. Scope of the standard library(What is covered by the Standard Library Program)?
> Everything that's often needed, i dont mind a big standard library.
12. Scope of RC, Ownership(When can I not use them?)?
> You should always use RC/ownership. The non-RC structs are just for people who know what they are doing, less overhead & it allows you to do hacky things.
13. Hidden control flow?
> We dont want any magic functions in the language with the exception of a few builtin functions e.g. `__before_free`. I guess macros are also a type of hidden control flow.
14. How is Map implemented?
> At the moment it's the most basic version a map. We will change it too a hash map soon.
15. Hash Map?
> Soon
16. Pattern matching?
> No ideas yet, we'll probably need to add regexes to the standard lib
17. Function overloading? (I think either none, or use an explicit function overload like Odin's  https://odin-lang.org/docs/overview/#explicit-procedure-overloading)
> Im not convinced we should have function overloading
18. Can I set a function in a file to be private?
> Not yet, im still thinking if we need it, feel free to inform me why
19. Can I create a function or struc/class in function
> Not yet, but we plan to add this
20. for?
> probably not, `while` and `each` seems fine
21. What is the syntax of inline function?
> You mean telling the compiler to inline a function? right now `func name() void %inline { ... }`

**Comment by ctxcode** [comment]
1. borrowing from Jai/Odin syntax(Especially Odin)?
> we havent looked at their syntax yet, so i cant judge right now
2. Semicolon? (I know this is already a question under consideration)
> we'll probably stick with `;`. It makes it more clear where lines end.
3. let the function return two variables at the same time and assign values at the call?
> Would be nice, although LLVM doesnt support it, however, rust does it, so it must be possible. Might be a feature in the future, probably not in 1.0.
4. something like Zig's break / continue
> Breaking / continue'ing outer loops would indeed be useful, we'll add this to the TODO list, shouldnt be too much work.
5. Take inspiration from cppfront?
> Doesnt really look interesting honestly. Just a basic c++ transpiler.
6. Create an error set with multiple errors?
> You can already have multiple errors and in the error handler, your error variable also represent the set of possible errors.
> `myfunc() !! |err| { if(err == err.my_error){ .. } exit 1; };`
7. use CPP lib?
> Not right now, maybe later, people can always create their own c++ packages to make the linker compatible with c++ libraries. Anyway, c++ is a pain the ass, it's so garbage, but we'll look into it.
8. defer?
> Maybe, it's a nice feature for the people managing their own memory.
9. anonymous blocks?
> It's a nice feature, but low priority, you don't really need it, but it's useful yes.
10. The syntax of the declaration?
> Everyone has their own preferences, people dont care that much. Maybe we can do a poll for it someday.
11. A random thought of my own. sth like:
> No, i know jai does it, but i prefer to oldschool way (for now).
12. named and optional arguments?
> Named is on the TODO list, optional arguments already exist

**Comment by C-BJ** [comment]
Zig's break not only exits the outer loop, but also returns a value

**Reply by ctxcode** [reply]
in `ki` you can do this using @vs (value scope). eg:
```
let x = @vs : i32 {
  while... {
    if ... return 15;
  }
  return 0;
};
```

**Reply by C-BJ** [reply]
great

**Reply by ctxcode** [reply]
maybe we can reduce it to `let x = { ... code ... };`

**Reply by C-BJ** [reply]
maybe, but in this case we should use break instead of return to return a value

**Comment by C-BJ** [comment]
I think the semicolon should be optional, which increases writability without reducing readability. I think it should be possible to set what content in the file is private, because you certainly don't want some internal classes to be public, right?

**Reply by ctxcode** [reply]
We will allow you to make classes/functions private. Side note: internal classes being public is actually not that bad, some people even prefer it.

**Comment by C-BJ** [comment]
Regarding Hash Map, I recommend you to check out the CppCon presentation, which presents a new extremely fast implementation of Hash Map

**Reply by ctxcode** [reply]
Hash maps will always be a trade off between memory, cpu and use cases, these algorithms they show at ccpcon are nothing new at all.
